### PR TITLE
Rework oriented containers

### PR DESCRIPTION
**Pull request by KoBeWi**
This PR makes BoxContainer, SplitContainer and FlowContainer non-abstract, so you can change the orientation on the fly using a `vertical` property:
godot windows tools 64_pYlwvVURZs

The property is not available in the specialized classes and trying to change it from code in these classes results in an error:
image
image

Related to https://github.com/godotengine/godot-proposals/issues/3558
Helps #36040

### CODE REVIEW

**Comment by YuriSizov** [comment]
Love the icons :D

File: doc/classes/BoxContainer.xml [hunk 1]
```diff
@@ -22,13 +22,25 @@
...
 			The alignment of the container's children (must be one of [constant ALIGNMENT_BEGIN], [constant ALIGNMENT_CENTER], or [constant ALIGNMENT_END]).
 		</member>
+		<member name="vertical" type="bool" setter="set_vertical" getter="is_vertical" default="false">
+			If [code]true[/code], the [BoxContainer] will arrange its children vertically, rather than horizontally.
+			Can't be changed when using [HBoxContainer] and [VBoxContainer].
+		</member>
 	</members>
 	<constants>
 		<constant name="ALIGNMENT_BEGIN" value="0" enum="AlignmentMode">
+			The child controls will be arranged at the beginning of the container (left or top, depending if vertical).
```
Comment (bruvzg) [inline doc/classes/BoxContainer.xml:32]: > "left or top"

It might be right if UI layout is right-to-left.

File: doc/classes/BoxContainer.xml [hunk 2]
```diff
@@ -22,13 +22,25 @@
...
 		</constant>
 		<constant name="ALIGNMENT_CENTER" value="1" enum="AlignmentMode">
+			The child controls will be centered in the container.
 		</constant>
 		<constant name="ALIGNMENT_END" value="2" enum="AlignmentMode">
+			The child controls will be arranged at the end of the container (right or bottom, depending if vertical).
```
Comment (bruvzg) [inline doc/classes/BoxContainer.xml:38]: Same here, it's left for RTL layout

**Review by bruvzg — APPROVED** [review]

**Comment by YeldhamDev** [comment]
I don't get it. Why keep the `H/VBoxContainer` nodes with this change?

**Comment by KoBeWi** [comment]
Because `H/VBoxContainer` have more clear usage than some "BoxContainer". Or at least that was the reasoning why unifying it was originally rejected.

**Comment by MewPurPur** [comment]
To me it's useful to be able to look in the SceneTree dock and see the orientation and how containers differ, rather than a soup of diagonal containers that don't say much. There is no way to change the icon based on a property atm, the technology isn't there yet

**Comment by YeldhamDev** [comment]
> Because `H/VBoxContainer` have more clear usage than some "BoxContainer". Or at least that was the reasoning why unifying it was originally rejected.

Then I just don't see the point of this PR. If anything, it would just make things confusing.

**Comment by KoBeWi** [comment]
The PR doesn't change much. Users can still use H/VBoxContainers, advanced users can use BoxContainer. The only difference is that BoxContainer is no longer grayed-out and you can click it. Maybe the intended usage could be better clarified in the docs, but other than that it's fine IMO.

File: scene/gui/box_container.h [hunk 3]
```diff
@@ -73,7 +79,7 @@ class HBoxContainer : public BoxContainer {
 
 public:
 	HBoxContainer() :
-			BoxContainer(false) {}
+			BoxContainer(false) { is_fixed = true; }
```
Comment (akien-mga) [inline scene/gui/box_container.h:82]:
```suggestion
			BoxContainer(false),
			is_fixed(true) {}
```
Same for others.

File: scene/gui/flow_container.h [hunk 4]
```diff
@@ -45,13 +45,19 @@ class FlowContainer : public Container {
 	void _resort();
 
 protected:
+	bool is_fixed = false;
+
 	void _notification(int p_what);
+	void _validate_property(PropertyInfo &property) const override;
```
Comment (akien-mga) [inline scene/gui/flow_container.h:51]:
```suggestion
	void _validate_property(PropertyInfo &p_property) const override;
```

**Review by akien-mga — APPROVED** [review]
Looks good to me, up to others in @godotengine/gui-nodes to approve the concept.

File: scene/gui/box_container.h (hunk 3 above)
Comment (KoBeWi) [reply scene/gui/box_container.h:82]: Does not work, I get `C2614 illegal member initialization: 'is_fixed' is not a base or member` error :/

Comment (akien-mga) [reply scene/gui/box_container.h:82]: Ah that's weird. I guess it's fine as is then.

**Comment by akien-mga** [comment]
Thanks!

**Comment by sairam4123** [comment]
What about other H/V classes?

**Comment by KoBeWi** [comment]
They are not container.

**Pull request merged**